### PR TITLE
Feature/df 2283  getaux modernisation

### PR DESCRIPTION
**Pull request by emina-str**

### CODE REVIEW

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/config/AvamRavImporterBatchConfig.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+                .writer(jdbcBatchItemWriter(this.lamdaDataSource, AUX_RAVADR_INSERT_QUERY, ravAddressImportParameter()))
+                .build();
+    }
+
+    private Step loadRavPBStep() {
+        return this.stepBuilderFactory.get("load-rav-pb-step")
+                .<WSRavSachbearbeiter, RavPB>chunk(this.avamRavImporterBatchProperties.getChunkSize())
+                .reader(new RavPBSOAPReader())
+                .processor((ItemProcessor<WSRavSachbearbeiter, RavPB>) item -> AvamRavAssembler.toRavPB(item))
+                .writer(jdbcBatchItemWriter(this.lamdaDataSource, AUX_RAV_SACHB_INSERT_QUERY, ravPBImportParameter()))
```
Comment (fabiangotzen) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/config/AvamRavImporterBatchConfig.java:114]: Lamda's could be replaced with Method References

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/avam/AvamException.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package ch.admin.seco.alv.basis.avam;
+
+public class AvamException extends RuntimeException {
+
+    private final String id;
+    private final String action;
+    private final String avamResponse;
```
Comment (fabiangotzen) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/avam/AvamException.java:7]: Getters or convert to local variables?

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/avam/AvamWebServiceClientConfig.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Bean
+    public Jaxb2Marshaller marshaller() {
+        Jaxb2Marshaller marshaller = new Jaxb2Marshaller();
+        marshaller.setClassesToBeBound(
+                WSRavAdresse.class, WSRavAdresseArray.class, GetRavAdressen.class, GetRavAdressenResponse.class,
+                WSRavSachbearbeiter.class, WSRavSachbearbeiterArray.class, GetRavSachbearbeiter.class, GetRavSachbearbeiterResponse.class,
+                TCode.class, TCodeArray.class, GetCode.class, GetCodeResponse.class
+        );
+        //marshaller.setPackagesToScan("ch.admin.seco.alv.basis.avam.ws");
```
Comment (fabiangotzen) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/avam/AvamWebServiceClientConfig.java:62]: Can be removed?

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    private final JobBuilderFactory jobBuilderFactory;
+
+    private final StepBuilderFactory stepBuilderFactory;
+
+    private final DataSource lamdaDataSource;
+
+    private final AsalAddressImporterBatchProperties asalAddressImporterBatchProperties;
+
+    private int ix;
```
Comment (fabiangotzen) [inline alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java:47]: Maybe the naming could be improved somehow?

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/avam/AvamException.java (hunk 2 above)
Comment (emina-str) [reply alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/avam/AvamException.java:7]: added getters.

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java (hunk 4 above)
Comment (emina-str) [reply alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java:47]: renamed to `asalAddressIxStartNumber`

File: alv-asal-address-importer/pom.xml [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>${liquibase.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
```
Comment (birom) [inline alv-asal-address-importer/pom.xml:48]: Is this for the ``AsalAddressImporterBatchProperties`` validation?

File: alv-asal-address-importer/pom.xml [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            <artifactId>hibernate-validator</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
```
Comment (birom) [inline alv-asal-address-importer/pom.xml:52]: We are not using ``jpa`` . I think we do not need this.

File: alv-asal-address-importer/pom.xml [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                    <execution>
+                        <id>build-info</id>
+                        <goals>
+                            <goal>build-info</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>pl.project13.maven</groupId>
```
Comment (birom) [inline alv-asal-address-importer/pom.xml:91]: What is this used for?

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        writer.setDataSource(this.lamdaDataSource);
+        writer.setAssertUpdates(true);
+        writer.setSql(AUX_ASALADR_INSERT_QUERY);
+        writer.setItemPreparedStatementSetter(parameter);
+        writer.afterPropertiesSet();
+
+        return writer;
+    }
+
+    private ItemPreparedStatementSetter<AsalAddress> asalAddressImportParameter() {
```
Comment (birom) [inline alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java:136]: IMHO, named parameters would be easier to maintain. 
As far as I remember, in that case the ``ItemPreparedStatementSetter`` can replaced with ```
writer.setItemSqlParameterSourceProvider(new BeanPropertyItemSqlParameterSourceProvider<T>());
```

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
+package ch.admin.seco.alv.basis.domain;
+
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.util.Collections;
+
+@Entity
```
Comment (birom) [inline alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java:8]: Please remove the ``JPA`` annotations.

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+    public static final String AUX_ASALADR = "AUX_ASALADR";
+    public static final String[] AUX_ASALADR_COLUMNS = new String[] {
+            "id", "vorname", "nachname", "adr1", "strasse", "plz", "ort", "telefon", "telefax", "funktion", "bereich", "email", "modus"
+    };
+    public static final String AUX_ASALADR_INSERT_QUERY = String.format(
+            "INSERT INTO %s (ix, %s, egov_zeit) VALUES (?, %s, now());",
+            AUX_ASALADR,
+            String.join(", ", AUX_ASALADR_COLUMNS),
+            String.join(", ", Collections.nCopies(AUX_ASALADR_COLUMNS.length, "?"))
```
Comment (birom) [inline alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java:20]: IMHO, named parameters would be easier to maintain.

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/config/AvamRavImporterBatchConfig.java [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+        StringBuilder sb = new StringBuilder();
+        for (String s : TABLES) {
+            sb.append(String.format(TRUNCATE_TABLE_SQL, s));
+        }
+        return sb.toString();
+    }
+
+    private ItemPreparedStatementSetter<RavAddress> ravAddressImportParameter() {
+        return (item, ps) -> {
+            ps.setString(1,  item.getCode());
```
Comment (birom) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/config/AvamRavImporterBatchConfig.java:222]: I think named parameters are easier to maintain.

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavAddress.java [hunk 12]
```diff
@@ -0,0 +1,217 @@
+package ch.admin.seco.alv.basis.domain;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.util.Collections;
+
+@Entity
+@Table(name = "AUX_RAVADR")
```
Comment (birom) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavAddress.java:10]: Please remove unused ``JPA`` annotations.

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPB.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
+package ch.admin.seco.alv.basis.domain;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.util.Collections;
+
+@Entity
+@Table(name = "AUX_RAV_SACHB")
```
Comment (birom) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPB.java:10]: Please remove ``JPA`` annotations.

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenter.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
+package ch.admin.seco.alv.basis.domain;
+
+import javax.persistence.Column;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import javax.persistence.Table;
+import java.util.Collections;
+
+@Entity
```
Comment (birom) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenter.java:9]: Please remove ``JPA`` annotations.

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenterId.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
+package ch.admin.seco.alv.basis.domain;
+
+import javax.persistence.Column;
+import javax.persistence.Embeddable;
+import java.io.Serializable;
+
+@Embeddable
```
Comment (birom) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenterId.java:7]: Please remove ``JPA`` annotations.

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/TCode.java [hunk 16]
```diff
@@ -0,0 +1,104 @@
+package ch.admin.seco.alv.basis.domain;
+
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.util.Collections;
+
+@Entity
+@Table(name = "T_CODE")
```
Comment (birom) [inline alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/TCode.java:9]: Please remove ``JPA`` annotations.

File: alv-asal-address-importer/pom.xml (hunk 5 above)
Comment (emina-str) [reply alv-asal-address-importer/pom.xml:48]: yes.

File: alv-asal-address-importer/pom.xml (hunk 6 above)
Comment (emina-str) [reply alv-asal-address-importer/pom.xml:52]: replaced with `javax.persistence`.

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/TCode.java (hunk 16 above)
Comment (emina-str) [reply alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/TCode.java:9]: done

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenterId.java (hunk 15 above)
Comment (emina-str) [reply alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenterId.java:7]: done

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenter.java (hunk 14 above)
Comment (emina-str) [reply alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPBJobCenter.java:9]: done

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPB.java (hunk 13 above)
Comment (emina-str) [reply alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavPB.java:10]: done

File: alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavAddress.java (hunk 12 above)
Comment (emina-str) [reply alv-avam-rav-importer/src/main/java/ch/admin/seco/alv/basis/domain/RavAddress.java:10]: done

File: alv-avam-rav-importer/pom.xml [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.ws.commons.axiom</groupId>
+            <artifactId>axiom-impl</artifactId>
+            <version>${ws.comon.axiom}</version>
+        </dependency>
+
+        <!-- Misc -->
+        <dependency>
+            <groupId>org.apache.commons</groupId>
```
[the remark this answers is not in the document]
Comment (emina-str) [reply alv-avam-rav-importer/pom.xml:103]: it is in the `marshaller()` Bean definition.

File: alv-avam-rav-importer/pom.xml [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+        </dependency>
+
+
+        <!-- Spring Cloud -->
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-starter-task</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
```
[the remark this answers is not in the document]
Comment (emina-str) [reply alv-avam-rav-importer/pom.xml:37]: yes, you're right. removed

File: alv-avam-rav-importer/pom.xml [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>${liquibase.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
```
[the remark this answers is not in the document]
Comment (emina-str) [reply alv-avam-rav-importer/pom.xml:119]: yes, for `properties`

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java (hunk 9 above)
Comment (emina-str) [reply alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java:8]: done

File: alv-asal-address-importer/pom.xml (hunk 6 above)
Comment (emina-str) [reply alv-asal-address-importer/pom.xml:52]: removed

File: alv-asal-address-importer/pom.xml (hunk 7 above)
Comment (emina-str) [reply alv-asal-address-importer/pom.xml:91]: Removed in both poms, also other plugins for docker. They were a try for the failing build, but it didn't help.

File: alv-avam-rav-importer/pom.xml [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+            <artifactId>hibernate-validator</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
```
[the remark this answers is not in the document]
Comment (emina-str) [reply alv-avam-rav-importer/pom.xml:123]: removed

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java (hunk 8 above)
Comment (emina-str) [reply alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java:136]: Added `named-params` for all tables in AvamRavImporter.

For AsalAddressImporter I have a problem with resolving field `IX` : now I'm setting it like 
`ps.setInt(1, this.asalAddressIxStartNumber++);` and it's working OK. To be able to use `named-params`, I tried adding in liquibase auto-increment option for `IX`, but it only worked when starting from 1 (even with `startWith=50 000`). Other option, adding it in `writer.setSql()`, I didn't find a working solution.

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java (hunk 10 above)
Comment (emina-str) [reply alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/domain/AsalAddress.java:15]: Added `named-params` for all tables in AvamRavImporter, but for `AsalAddress` I didn't find a way.

File: alv-asal-address-importer/pom.xml (hunk 7 above)
Comment (seco-mgabor) [reply alv-asal-address-importer/pom.xml]: This is used to display build information by the Spring actuator, but for batch jobs we didn't use it anywhere.

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+                .tasklet((contribution, chunkContext) -> {
+                    this.asalAddressIxStartNumber = this.asalAddressImporterBatchProperties.getTableIdStartWith();
+                    return RepeatStatus.FINISHED;
+                })
+                .build();
+    }
+
+    @Bean
+    public FlatFileItemReader<AsalAddress> flatFileItemReader() {
+        FlatFileItemReader<AsalAddress> reader = new FlatFileItemReader<>();
```
Comment (seco-mgabor) [inline alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfig.java:102]: What are you trying to tell us? :)

File: alv-asal-address-importer/src/test/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfigTest.java [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+import static ch.admin.seco.alv.basis.domain.AsalAddress.AUX_ASALADR;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.springframework.batch.core.BatchStatus.COMPLETED;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(properties = {
+        "spring.batch.job.enabled=false"
+})
+@ActiveProfiles("test")
+@Sql(
```
Comment (seco-mgabor) [inline alv-asal-address-importer/src/test/java/ch/admin/seco/alv/basis/config/AsalAddressImporterBatchConfigTest.java:33]: Why "lamda"? Isn't it something about addresses?

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/asaladdress/importer/SftpConfig.java [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+public class SftpConfig {
+    @Bean
+    public DefaultSftpSessionFactory sftpSessionFactory(SftpProperties sftpProperties) {
+        DefaultSftpSessionFactory factory = new DefaultSftpSessionFactory();
+        factory.setHost(sftpProperties.getHost());
+        if (nonNull(sftpProperties.getPort())) {
+            factory.setPort(sftpProperties.getPort());
+        }
+        if (nonNull(sftpProperties.getProxyHost())) {
+            int proxyPort = defaultIfNull(sftpProperties.getProxyPort(), 80);
```
Comment (fabiangotzen) [inline alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/asaladdress/importer/SftpConfig.java:29]: Magic number

File: alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/asaladdress/importer/AsalAddressImporterBatchConfig.java [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+
+    private JobExecutionListener cleanUpListener() {
+        return new JobExecutionListenerSupport() {
+            @Override
+            public void afterJob(JobExecution jobExecution) {
+
+                if (jobExecution.getExecutionContext().containsKey(PARAMETER_FLAT_FILE_PATH)) {
+                    File flatFile = (File) (jobExecution.getExecutionContext().get(PARAMETER_FLAT_FILE_PATH));
+                    try {
+                        Files.delete(flatFile.toPath());
```
Comment (fabiangotzen) [inline alv-asal-address-importer/src/main/java/ch/admin/seco/alv/basis/asaladdress/importer/AsalAddressImporterBatchConfig.java:219]: Guess not *really* a problem, but possible nullpointer.

**Pull request merged**